packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -992,6 +1002,39 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         this.updateDocumentDirtyState(newState);
     }
 
+    /**
+     * Used to rebase pending initial ops at their reference sequence number.
+     * Normal op processing is synchronous, but rebasing is async since the
+     * data store may not be loaded yet, so we pause DeltaManager between ops.
+     * It's also important that we see each op so we know all pending ops have
+     * been rebased by "connected" event, but process() doesn't see system ops,
+     * so we listen directly from DeltaManager instead.
+     */
+    private readonly onOp = (op: ISequencedDocumentMessage) => {
+        assert(!this.paused);
+        this.paused = true;
+        this.scheduleManager.setPaused(true);
+        const rebaseP = this.pendingStateManager.rebaseAt(op.sequenceNumber);
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        rebaseP.then(() => {
```

Comment:
We should account for exceptions here. I think right thing to do if we hit exception - close container with error